File: csharp/Dson/DsonObjectReader.cs
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections;
using Dson.Collections;
using Dson.IO;
using Dson.Types;
using Google.Protobuf;

namespace Dson;

public class DsonObjectReader<TName> : AbstractDsonReader<TName> where TName : IEquatable<TName>
{
    private TName? _nextName;
    private DsonValue? _nextValue;

    public DsonObjectReader(DsonReaderSettings settings, DsonArray<TName> dsonArray)
        : base(settings) {
        Context context = new Context();
        context.init(null, DsonContextType.TOP_LEVEL, DsonTypes.INVALID);
        context.ArrayIterator = new MarkableIterator<DsonValue>(dsonArray.GetEnumerator());
        SetContext(context);
    }

    /// <summary>
    /// 设置key的迭代顺序
    /// </summary>
    /// <param name="keyItr">Key的迭代器</param>
    /// <param name="defValue">key不存在时的返回值</param>
    public void SetKeyItr(IEnumerator<TName> keyItr, DsonValue defValue) {
        if (keyItr == null) throw new ArgumentNullException(nameof(keyItr));
        if (defValue == null) throw new ArgumentNullException(nameof(defValue));
        Context context = GetContext();
        if (context.DsonObject == null) {
            throw DsonIOException.contextError(DsonInternals.NewList(DsonContextType.OBJECT, DsonContextType.HEADER), context.contextType);
        }
        context.SetKeyItr(keyItr, defValue);
    }

    /// <summary>
    /// 获取当前对象的所有key
    /// </summary>
    /// <returns></returns>
    /// <exception cref="DsonIOException"></exception>
    public ICollection<TName> Keys() {
        Context context = GetContext();
        if (context.DsonObject == null) {
            throw DsonIOException.contextError(DsonInternals.NewList(DsonContextType.OBJECT, DsonContextType.HEADER), context.contextType);
        }
        return context.DsonObject.Keys;
    }

    protected override Context GetContext() {
        return (Context)_context;
    }

    protected override Context? GetPooledContext() {
        return (Context)base.GetPooledContext();
    }

    #region state

    private void PushNextValue(DsonValue nextValue) {
        if (nextValue == null) throw new ArgumentNullException(nameof(nextValue));
        this._nextValue = nextValue;
    }

    private DsonValue PopNextValue() {
        DsonValue r = this._nextValue;
        this._nextValue = null;
        return r;
    }

    private void PushNextName(TName nextName) {
        this._nextName = nextName;
    }

    private TName PopNextName() {
        TName r = this._nextName;
        this._nextName = default;
        return r;
    }

    public override DsonType ReadDsonType() {
        Context context = GetContext();
        checkReadDsonTypeState(context);

        PopNextName();
        PopNextValue();

        DsonType dsonType;
        if (context.Header != null) { // 需要先读取header
            dsonType = DsonType.HEADER;
            PushNextValue(context.Header);
            context.Header = null;
        }
        else if (context.contextType.isLikeArray()) {
            DsonValue nextValue = context.NextValue();
            if (nextValue == null) {
                dsonType = DsonType.END_OF_OBJECT;
            }
            else {
                PushNextValue(nextValue);
                dsonType = nextValue.DsonType;
            }
        }
        else {
            KeyValuePair<TName, DsonValue>? nextElement = context.NextElement();
            if (!nextElement.HasValue) {
                dsonType = DsonType.END_OF_OBJECT;
            }
            else {
                PushNextName(nextElement.Value.Key);
                PushNextValue(nextElement.Value.Value);
                dsonType = nextElement.Value.Value.DsonType;
            }
        }

        this.currentDsonType = dsonType;
        this.currentWireType = WireType.VarInt;
        this.currentName = default;

        onReadDsonType(context, dsonType);
        return dsonType;
    }

    public override DsonType PeekDsonType() {
        Context context = this.GetContext();
        checkReadDsonTypeState(context);

        if (context.Header != null) {
            return DsonType.HEADER;
        }
        if (!context.HasNext()) {
            return DsonType.END_OF_OBJECT;
        }
        if (context.contextType.isLikeArray()) {
            context.MarkItr();
            DsonValue nextValue = context.NextValue();
            context.ResetItr();
            return nextValue!.DsonType;
        }
        else {
            context.MarkItr();
            KeyValuePair<TName, DsonValue>? nextElement = context.NextElement();
            context.ResetItr();
            return nextElement!.Value.Value.DsonType;
        }
    }

    protected override void doReadName() {
        currentName = PopNextName();
    }

    #endregion

    #region 简单值

    protected override int doReadInt32() {
        return PopNextValue().AsInt32();
    }

    protected override long doReadInt64() {
        return PopNextValue().AsInt64();
    }

    protected override float doReadFloat() {
        return PopNextValue().AsFloat();
    }

    protected override double doReadDouble() {
        return PopNextValue().AsDouble();
    }

    protected override bool doReadBool() {
        return PopNextValue().AsBool();
    }

    protected override string doReadString() {
        return PopNextValue().AsString();
    }

    protected override void doReadNull() {
        PopNextValue();
    }

    protected override DsonBinary doReadBinary() {
        return PopNextValue().AsBinary().Copy(); // 需要拷贝
    }

    protected override DsonExtInt32 doReadExtInt32() {
        return PopNextValue().AsExtInt32();
    }

    protected override DsonExtInt64 doReadExtInt64() {
        return PopNextValue().AsExtInt64();
    }

    protected override DsonExtDouble doReadExtDouble() {
        return PopNextValue().AsExtDouble();
    }

    protected override DsonExtString doReadExtString() {
        return PopNextValue().AsExtString();
    }

    protected override ObjectRef doReadRef() {
        return PopNextValue().AsReference();
    }

    protected override OffsetTimestamp doReadTimestamp() {
        return PopNextValue().AsTimestamp();
    }

    #endregion

    #region 容器

    protected override void doReadStartContainer(DsonContextType contextType, DsonType dsonType) {
        Context newContext = NewContext(GetContext(), contextType, dsonType);
        DsonValue dsonValue = PopNextValue();
        if (dsonValue.DsonType == DsonType.OBJECT) {
            DsonObject<TName> dsonObject = dsonValue.AsObject<TName>();
            newContext.Header = dsonObject.Header.Count > 0 ? dsonObject.Header : null;
            newContext.DsonObject = dsonObject;
            newContext.ObjectIterator = new MarkableIterator<KeyValuePair<TName, DsonValue>>(dsonObject.GetEnumerator());
        }
        else if (dsonValue.DsonType == DsonType.ARRAY) {
            DsonArray<TName> dsonArray = dsonValue.AsArray<TName>();
            newContext.Header = dsonArray.Header.Count > 0 ? dsonArray.Header : null;
            newContext.ArrayIterator = new MarkableIterator<DsonValue>(dsonArray.GetEnumerator());
        }
        else {
            // 其它内置结构体
            newContext.DsonObject = dsonValue.AsHeader<TName>();
            newContext.ObjectIterator = new MarkableIterator<KeyValuePair<TName, DsonValue>>(dsonValue.AsHeader<TName>().GetEnumerator());
        }
        newContext.name = currentName;

        this.recursionDepth++;
        SetContext(newContext);
    }

    protected override void doReadEndContainer() {
        Context context = GetContext();

        // 恢复上下文
        recoverDsonType(context);
        this.recursionDepth--;
        SetContext(context.parent!);
        PoolContext(context);
    }

    #endregion

    #region 特殊

    protected override void doSkipName() {
        PopNextName();
    }


    protected override void doSkipValue() {
        PopNextValue();
    }

    protected override void doSkipToEndOfObject() {
        Context context = GetContext();
        context.Header = null;
        if (context.ArrayIterator != null) {
            context.ArrayIterator.ForEachRemaining(e => { });
        }
        else {
            context.ObjectIterator!.ForEachRemaining(e => { });
        }
    }

    protected override T doReadMessage<T>(int binaryType, MessageParser<T> parser) {
        DsonBinary dsonBinary = (DsonBinary)PopNextValue();
        if (dsonBinary == null) throw new InvalidOperationException();
        return parser.ParseFrom(dsonBinary.Data);
    }


    protected override byte[] doReadValueAsBytes() {
        throw new InvalidOperationException("Unsupported operation");
    }

    #endregion

    #region context

    private Context NewContext(Context parent, DsonContextType contextType, DsonType dsonType) {
        Context context = GetPooledContext();
        if (context != null) {
            SetPooledContext(null);
        }
        else {
            context = new Context();
        }
        context.init(parent, contextType, dsonType);
        return context;
    }

    private void PoolContext(Context context) {
        context.reset();
        SetPooledContext(context);
    }

    protected new class Context : AbstractDsonReader<TName>.Context
    {
        /** 如果不为null，则表示需要先读取header */
        protected internal DsonHeader<TName>? Header;
        protected internal AbstractDsonObject<TName>? DsonObject;
        protected internal MarkableIterator<KeyValuePair<TName, DsonValue>>? ObjectIterator;
        protected internal MarkableIterator<DsonValue>? ArrayIterator;

        public Context() {
        }

        public override void reset() {
            base.reset();
            Header = null;
            DsonObject = null;
            ObjectIterator = null;
            ArrayIterator = null;
        }

        public void SetKeyItr(IEnumerator<TName> keyItr, DsonValue defValue) {
            if (DsonObject == null) throw new InvalidOperationException();
            if (ObjectIterator!.IsMarking) throw new InvalidOperationException("reader is in marking state");

            ObjectIterator = new MarkableIterator<KeyValuePair<TName, DsonValue>>(new KeyIterator(DsonObject, keyItr, defValue));
        }

        public bool HasNext() {
            if (ObjectIterator != null) {
                return ObjectIterator.HasNext();
            }
            else {
                return ArrayIterator!.HasNext();
            }
        }

        public void MarkItr() {
            if (ObjectIterator != null) {
                ObjectIterator.Mark();
            }
            else {
                ArrayIterator!.Mark();
            }
        }

        public void ResetItr() {
            if (ObjectIterator != null) {
                ObjectIterator.Reset();
            }
            else {
                ArrayIterator!.Reset();
            }
        }

        public DsonValue? NextValue() {
            return ArrayIterator!.HasNext() ? ArrayIterator.Next() : null;
        }

        public KeyValuePair<TName, DsonValue>? NextElement() {
            return ObjectIterator!.HasNext() ? ObjectIterator.Next() : null;
        }
    }

    private class KeyIterator : IEnumerator<KeyValuePair<TName, DsonValue>>
    {
        private readonly AbstractDsonObject<TName> _dsonObject;
        private readonly IEnumerator<TName> _keyItr;
        private readonly DsonValue _defValue;

        public KeyIterator(AbstractDsonObject<TName> dsonObject, IEnumerator<TName> keyItr, DsonValue defValue) {
            this._dsonObject = dsonObject;
            this._keyItr = new MarkableIterator<TName>(keyItr);
            this._defValue = defValue;
        }

        public bool MoveNext() {
            return _keyItr.MoveNext();
        }

        public KeyValuePair<TName, DsonValue> Current {
            get {
                TName key = _keyItr.Current;
                if (_dsonObject.TryGetValue(key!, out DsonValue dsonValue)) {
                    return new KeyValuePair<TName, DsonValue>(key, dsonValue);
                }
                else {
                    return new KeyValuePair<TName, DsonValue>(key, _defValue);
                }
            }
        }

        public void Reset() {
            _keyItr.Reset();
        }

        object IEnumerator.Current => Current;

        public void Dispose() {
        }
    }

    #endregion
}

File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonObjectReader.java
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import cn.wjybxx.dson.internal.MarkableIterator;
import cn.wjybxx.dson.io.DsonIOException;
import cn.wjybxx.dson.types.ObjectRef;
import cn.wjybxx.dson.types.OffsetTimestamp;

import java.util.*;

/**
 * @author wjybxx
 * date - 2023/6/13
 */
public class DsonObjectReader extends AbstractDsonReader {

    private String nextName;
    private DsonValue nextValue;

    public DsonObjectReader(DsonReaderSettings settings, DsonArray<String> dsonArray) {
        super(settings);
        Context context = new Context();
        context.init(null, DsonContextType.TOP_LEVEL, null);
        context.header = dsonArray.getHeader().size() > 0 ? dsonArray.getHeader() : null;
        context.arrayIterator = new MarkableIterator<>(dsonArray.iterator());
        setContext(context);
    }

    @Override
    protected Context getContext() {
        return (Context) super.getContext();
    }

    @Override
    protected Context getPooledContext() {
        return (Context) super.getPooledContext();
    }

    /**
     * 设置key的迭代顺序
     *
     * @param defValue key不存在时的返回值；可选择{@link DsonNull#UNDEFINE}
     */
    public void setKeyItr(Iterator<String> keyItr, DsonValue defValue) {
        Objects.requireNonNull(keyItr);
        Objects.requireNonNull(defValue);
        Context context = getContext();
        if (context.dsonObject == null) {
            throw DsonIOException.contextError(List.of(DsonContextType.OBJECT, DsonContextType.HEADER), context.contextType);
        }
        context.setKeyItr(keyItr, defValue);
    }

    public Set<String> getkeySet() {
        Context context = getContext();
        if (context.dsonObject == null) {
            throw DsonIOException.contextError(List.of(DsonContextType.OBJECT, DsonContextType.HEADER), context.contextType);
        }
        return context.dsonObject.keySet();
    }

    // region state

    private void pushNextValue(DsonValue nextValue) {
        this.nextValue = Objects.requireNonNull(nextValue);
    }

    private DsonValue popNextValue() {
        DsonValue r = this.nextValue;
        this.nextValue = null;
        return r;
    }

    private void pushNextName(String nextName) {
        this.nextName = Objects.requireNonNull(nextName);
    }

    private String popNextName() {
        String r = this.nextName;
        this.nextName = null;
        return r;
    }

    @Override
    public DsonType readDsonType() {
        Context context = this.getContext();
        checkReadDsonTypeState(context);

        popNextName();
        popNextValue();

        DsonType dsonType;
        if (context.header != null) { // 需要先读取header
            dsonType = DsonType.HEADER;
            pushNextValue(context.header);
            context.header = null;
        } else if (context.contextType.isLikeArray()) {
            DsonValue nextValue = context.nextValue();
            if (nextValue == null) {
                dsonType = DsonType.END_OF_OBJECT;
            } else {
                pushNextValue(nextValue);
                dsonType = nextValue.getDsonType();
            }
        } else {
            Map.Entry<String, DsonValue> nextElement = context.nextElement();
            if (nextElement == null) {
                dsonType = DsonType.END_OF_OBJECT;
            } else {
                pushNextName(nextElement.getKey());
                pushNextValue(nextElement.getValue());
                dsonType = nextElement.getValue().getDsonType();
            }
        }

        this.currentDsonType = dsonType;
        this.currentWireType = WireType.VARINT;
        this.currentName = INVALID_NAME;

        onReadDsonType(context, dsonType);
        return dsonType;
    }

    @Override
    public DsonType peekDsonType() {
        Context context = this.getContext();
        checkReadDsonTypeState(context);

        if (context.header != null) {
            return DsonType.HEADER;
        }
        if (!context.hasNext()) {
            return DsonType.END_OF_OBJECT;
        }
        if (context.contextType.isLikeArray()) {
            context.markItr();
            DsonValue nextValue = context.nextValue();
            context.resetItr();
            return nextValue.getDsonType();
        } else {
            context.markItr();
            Map.Entry<String, DsonValue> nextElement = context.nextElement();
            context.resetItr();
            return nextElement.getValue().getDsonType();
        }
    }

    @Override
    protected void doReadName() {
        currentName = popNextName();
    }

    // endregion

    // region 简单值

    @Override
    protected int doReadInt32() {
        return popNextValue().asInt32(); // as顺带null检查
    }

    @Override
    protected long doReadInt64() {
        return popNextValue().asInt64();
    }

    @Override
    protected float doReadFloat() {
        return popNextValue().asFloat();
    }

    @Override
    protected double doReadDouble() {
        return popNextValue().asDouble();
    }

    @Override
    protected boolean doReadBool() {
        return popNextValue().asBool();
    }

    @Override
    protected String doReadString() {
        return popNextValue().asString();
    }

    @Override
    protected void doReadNull() {
        popNextValue();
    }

    @Override
    protected DsonBinary doReadBinary() {
        return popNextValue().asBinary().copy(); // 需要拷贝
    }

    @Override
    protected DsonExtInt32 doReadExtInt32() {
        return popNextValue().asExtInt32();
    }

    @Override
    protected DsonExtInt64 doReadExtInt64() {
        return popNextValue().asExtInt64();
    }

    @Override
    protected DsonExtDouble doReadExtDouble() {
        return popNextValue().asExtDouble();
    }

    @Override
    protected DsonExtString doReadExtString() {
        return popNextValue().asExtString();
    }

    @Override
    protected ObjectRef doReadRef() {
        return popNextValue().asReference();
    }

    @Override
    protected OffsetTimestamp doReadTimestamp() {
        return popNextValue().asTimestamp();
    }

    // endregion

    // region 容器

    @Override
    protected void doReadStartContainer(DsonContextType contextType, DsonType dsonType) {
        Context newContext = newContext(getContext(), contextType, dsonType);
        DsonValue dsonValue = popNextValue();
        if (dsonValue.getDsonType() == DsonType.OBJECT) {
            DsonObject<String> dsonObject = dsonValue.asObject();
            newContext.header = dsonObject.getHeader().size() > 0 ? dsonObject.getHeader() : null;
            newContext.dsonObject = dsonObject;
            newContext.objectIterator = new MarkableIterator<>(dsonObject.entrySet().iterator());
        } else if (dsonValue.getDsonType() == DsonType.ARRAY) {
            DsonArray<String> dsonArray = dsonValue.asArray();
            newContext.header = dsonArray.getHeader().size() > 0 ? dsonArray.getHeader() : null;
            newContext.arrayIterator = new MarkableIterator<>(dsonArray.iterator());
        } else {
            // 其它内置结构体
            newContext.dsonObject = dsonValue.asHeader();
            newContext.objectIterator = new MarkableIterator<>(dsonValue.asHeader().entrySet().iterator());
        }
        newContext.name = currentName;

        this.recursionDepth++;
        setContext(newContext);
    }

    @Override
    protected void doReadEndContainer() {
        Context context = getContext();

        // 恢复上下文
        recoverDsonType(context);
        this.recursionDepth--;
        setContext(context.parent);
        poolContext(context);
    }

    // endregion

    // region 特殊接口

    @Override
    protected void doSkipName() {
        popNextName();
    }

    @Override
    protected void doSkipValue() {
        popNextValue();
    }

    @Override
    protected void doSkipToEndOfObject() {
        Context context = getContext();
        context.header = null;
        if (context.arrayIterator != null) {
            context.arrayIterator.forEachRemaining(dsonValue -> {});
        } else {
            context.objectIterator.forEachRemaining(e -> {});
        }
    }

    @Override
    protected byte[] doReadValueAsBytes() {
        throw new UnsupportedOperationException();
    }

    // endregion

    // region context

    private Context newContext(Context parent, DsonContextType contextType, DsonType dsonType) {
        Context context = getPooledContext();
        if (context != null) {
            setPooledContext(null);
        } else {
            context = new Context();
        }
        context.init(parent, contextType, dsonType);
        return context;
    }

    private void poolContext(Context context) {
        context.reset();
        setPooledContext(context);
    }

    protected static class Context extends AbstractDsonReader.Context {

        /** 如果不为null，则表示需要先读取header */
        private DsonHeader<String> header;
        private AbstractDsonObject<String> dsonObject;
        private MarkableIterator<Map.Entry<String, DsonValue>> objectIterator;
        private MarkableIterator<DsonValue> arrayIterator;

        public Context() {
        }

        public void reset() {
            super.reset();
            header = null;
            dsonObject = null;
            objectIterator = null;
            arrayIterator = null;
        }

        public void setKeyItr(Iterator<String> keyItr, DsonValue defValue) {
            if (dsonObject == null) throw new IllegalStateException();
            if (objectIterator.isMarking()) throw new IllegalStateException("reader is in marking state");
            objectIterator = new MarkableIterator<>(new KeyIterator(dsonObject, keyItr, defValue));
        }

        public boolean hasNext() {
            if (objectIterator != null) {
                return objectIterator.hasNext();
            } else {
                return arrayIterator.hasNext();
            }
        }

        public void markItr() {
            if (objectIterator != null) {
                objectIterator.mark();
            } else {
                arrayIterator.mark();
            }
        }

        public void resetItr() {
            if (objectIterator != null) {
                objectIterator.reset();
            } else {
                arrayIterator.reset();
            }
        }

        public DsonValue nextValue() {
            return arrayIterator.hasNext() ? arrayIterator.next() : null;
        }

        public Map.Entry<String, DsonValue> nextElement() {
            return objectIterator.hasNext() ? objectIterator.next() : null;
        }

    }

    private static class KeyIterator implements Iterator<Map.Entry<String, DsonValue>> {

        final AbstractDsonObject<String> dsonObject;
        final Iterator<String> keyItr;
        final DsonValue defValue;

        public KeyIterator(AbstractDsonObject<String> dsonObject, Iterator<String> keyItr, DsonValue defValue) {
            this.dsonObject = dsonObject;
            this.keyItr = keyItr;
            this.defValue = defValue;
        }

        @Override
        public boolean hasNext() {
            return keyItr.hasNext();
        }

        @Override
        public Map.Entry<String, DsonValue> next() {
            String key = keyItr.next();
            DsonValue dsonValue = dsonObject.get(key);
            if (dsonValue == null) {
                return Map.entry(key, defValue);
            } else {
                return Map.entry(key, dsonValue);
            }
        }
    }

    // endregion

}